File: plot_output.m
% Charger les données depuis le fichier (modifier 'data.txt' avec le vrai nom)
data = load('u_N_plusResu_k=100.000000_a=1.000000_N=150.txt');

% Extraire les colonnes
x = data(:, 1);
y = data(:, 2);
P_real= data(:,3);
P_imag = data(:,4);

% Extraction de la partie réelle de P depuis la 3ème colonne
% On suppose que la colonne 3 est sous la forme (Re, Im), donc on garde uniquement Re
%P_real = sscanf(data(:, 3), '(%f,%f)');  % Si stocké sous format texte "(Re, Im)"

% Créer une grille régulière pour interpolation
xq = linspace(min(x), max(x), 100);  % 100 points en X
yq = linspace(min(y), max(y), 100);  % 100 points en Y
[Xq, Yq] = meshgrid(xq, yq);

% Interpolation des valeurs de P_real sur la grille
Pq = griddata(x, y, P_real, Xq, Yq, 'linear');



% Affichage de la carte de chaleur
figure;
imagesc(xq, yq, Pq);
colorbar;
xlabel('X');
ylabel('Y');
axis equal;
title('Partie réelle de la pression P');
%set(gca, 'YDir', 'normal'); % Correction pour l'orientation de l'image
% Modifier la taille des axes et de la colorbar
set(gca, 'FontSize', 40);
set(colorbar, 'FontSize', 40);

hold on; % Superposition des cercles noirs

% Charger les cercles depuis maillage.txt
maillage_data = load('outputs/obstacles.txt');

% Déterminer si maillage.txt contient les centres et rayons des cercles
if size(maillage_data, 2) == 3
    % Format : (xc, yc, r) avec xc, yc centre et r rayon
    for i = 1:size(maillage_data, 1)
        xc = maillage_data(i, 1);
        yc = maillage_data(i, 2);
        r = maillage_data(i, 3);

        % Tracer et remplir le cercle en noir
        theta = linspace(0, 2*pi, 100); % 100 points pour un cercle lisse
        x_circle = xc + r * cos(theta);
        y_circle = yc + r * sin(theta);
        fill(x_circle, y_circle, 'k', 'EdgeColor', 'none'); % Cercle noir sans bord
    end
else
    disp("Le fichier maillage.txt ne contient pas des cercles sous la forme (xc, yc, r)");
endif




hold off; % Arrêter la superposition



% Interpolation des valeurs de P_imag sur la grille
Pq = griddata(x, y, P_imag, Xq, Yq, 'linear');



% Affichage de la carte de chaleur
figure;
imagesc(xq, yq, Pq);
colorbar;
xlabel('X');
ylabel('Y');
axis equal;
title('Partie imaginaire de la pression P');
%set(gca, 'YDir', 'normal'); % Correction pour l'orientation de l'image
% Modifier la taille des axes et de la colorbar
set(gca, 'FontSize', 40);
set(colorbar, 'FontSize', 40);

hold on; % Superposition des cercles noirs

% Charger les cercles depuis maillage.txt
maillage_data = load('outputs/obstacles.txt');

% Déterminer si maillage.txt contient les centres et rayons des cercles
if size(maillage_data, 2) == 3
    % Format : (xc, yc, r) avec xc, yc centre et r rayon
    for i = 1:size(maillage_data, 1)
        xc = maillage_data(i, 1);
        yc = maillage_data(i, 2);
        r = maillage_data(i, 3);

        % Tracer et remplir le cercle en noir
        theta = linspace(0, 2*pi, 100); % 100 points pour un cercle lisse
        x_circle = xc + r * cos(theta);
        y_circle = yc + r * sin(theta);
        fill(x_circle, y_circle, 'k', 'EdgeColor', 'none'); % Cercle noir sans bord
    end
else
    disp("Le fichier maillage.txt ne contient pas des cercles sous la forme (xc, yc, r)");
endif




hold off; % Arrêter la superposition


% Interpolation des valeurs de norme de P sur la grille
P_norm = sqrt(P_real.^2 + P_imag.^2);
Pq = griddata(x, y, P_norm, Xq, Yq, 'linear');



% Affichage de la carte de chaleur
figure;
imagesc(xq, yq, Pq);
colorbar;
xlabel('X');
ylabel('Y');
axis equal;
title('Norme de la pression complexe P');
%set(gca, 'YDir', 'normal'); % Correction pour l'orientation de l'image
% Modifier la taille des axes et de la colorbar
set(gca, 'FontSize', 40);
set(colorbar, 'FontSize', 40);

hold on; % Superposition des cercles noirs

% Charger les cercles depuis maillage.txt
maillage_data = load('outputs/obstacles.txt');

% Déterminer si maillage.txt contient les centres et rayons des cercles
if size(maillage_data, 2) == 3
    % Format : (xc, yc, r) avec xc, yc centre et r rayon
    for i = 1:size(maillage_data, 1)
        xc = maillage_data(i, 1);
        yc = maillage_data(i, 2);
        r = maillage_data(i, 3);

        % Tracer et remplir le cercle en noir
        theta = linspace(0, 2*pi, 100); % 100 points pour un cercle lisse
        x_circle = xc + r * cos(theta);
        y_circle = yc + r * sin(theta);
        fill(x_circle, y_circle, 'k', 'EdgeColor', 'none'); % Cercle noir sans bord
    end
else
    disp("Le fichier maillage.txt ne contient pas des cercles sous la forme (xc, yc, r)");
endif




hold off; % Arrêter la superposition

File: plot_output.py
import os

import matplotlib.pyplot as plt
import numpy as np
from matplotlib.patches import Circle
from scipy.interpolate import griddata

# 1. Charger les données depuis le fichier
data_file = 'outputs/u_N_plusResu_k=10.000000_a=1.000000_N=15.txt'
data = np.loadtxt(data_file)

# Nom de base des PNG : même nom que le fichier de données (contient k, a et N)
base_name = os.path.splitext(os.path.basename(data_file))[0]
save_dir = 'outputs'
os.makedirs(save_dir, exist_ok=True)

# Extraire les colonnes
x = data[:, 0]
y = data[:, 1]
P_real = data[:, 2]
P_imag = data[:, 3]

# Créer une grille régulière pour interpolation
xq = np.linspace(x.min(), x.max(), 100)
yq = np.linspace(y.min(), y.max(), 100)
Xq, Yq = np.meshgrid(xq, yq)

# Charger les obstacles et s'assurer qu'ils sont en 2 dimensions
maillage_data = np.loadtxt('outputs/obstacles.txt')
maillage_data = np.atleast_2d(maillage_data)


def plot_pressure_field(P_values, title_text, suffix):
  # Interpolation des valeurs sur la grille
  Pq = griddata((x, y), P_values, (Xq, Yq), method='linear')

  # Masquer (mettre à NaN) les valeurs à l'intérieur des cercles
  if maillage_data.shape[1] == 3:
    for i in range(maillage_data.shape[0]):
      xc = maillage_data[i, 0]
      yc = maillage_data[i, 1]
      r = maillage_data[i, 2]

      inside_circle = (Xq - xc) ** 2 + (Yq - yc) ** 2 <= r**2
      Pq[inside_circle] = np.nan

  fig, ax = plt.subplots(figsize=(8, 8))

  # Affichage de la carte de chaleur
  extent = [xq.min(), xq.max(), yq.min(), yq.max()]
  im = ax.imshow(
      Pq, extent=extent, origin='lower', aspect='equal', cmap='viridis'
  )

  # Colorbar et axes
  cbar = fig.colorbar(im, ax=ax)
  cbar.ax.tick_params(labelsize=20)
  ax.set_xlabel('X', fontsize=20)
  ax.set_ylabel('Y', fontsize=20)
  ax.set_title(title_text, fontsize=20)
  ax.tick_params(labelsize=20)

  # Superposition des cercles noirs pleins
  if maillage_data.shape[1] == 3:
    for i in range(maillage_data.shape[0]):
      xc = maillage_data[i, 0]
      yc = maillage_data[i, 1]
      r = maillage_data[i, 2]

      circle = Circle((xc, yc), r, color='k', fill=True)
      ax.add_patch(circle)
  else:
    print(
        'Le fichier maillage.txt ne contient pas des cercles sous la forme (xc,'
        ' yc, r)'
    )  #

  # Sauvegarde du PNG (avant plt.show(), sinon la figure est vidée à la fermeture)
  png_path = os.path.join(save_dir, f'{base_name}_{suffix}.png')
  fig.savefig(png_path, dpi=200, bbox_inches='tight')
  print('saved', png_path)

  plt.show()


# --- 1. Partie réelle de la pression P ---
plot_pressure_field(P_real, 'Partie réelle de la solution exterieure u+', 'real')

# --- 2. Partie imaginaire de la pression P ---
plot_pressure_field(P_imag, 'Partie imaginaire de la solution exterieure u+', 'imag')

# --- 3. Norme de la pression complexe P ---
P_norm = np.sqrt(P_real**2 + P_imag**2)
plot_pressure_field(P_norm, 'Norme de la solution exterieure complexe u+', 'norm')

